File: csharp/Assembler/Common/Logger.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Assembler.Common
{
    /// <summary>
    /// Simple custom logger for consistent logging across all language implementations.
    /// Provides file-based logging with different log levels.
    /// </summary>
    public static class Logger
    {
        public enum LogLevel
        {
            DEBUG = 0,
            INFO = 1,
            WARN = 2,
            ERROR = 3,
            NONE = 4
        }

        public enum LogRotation
        {
            NONE = 0,
            HOURLY = 1,
            DAILY = 2
        }

        private static LogLevel _currentLogLevel = LogLevel.NONE;
        private static string? _logFilePath = null;
        private static bool _consoleOutput = true;
        private static readonly object _lock = new object();
        private static LogRotation _logRotation = LogRotation.NONE;
        private static string? _baseLogFilePath = null;
        private static string? _currentRotatedPath = null;

        // Support for context-specific log files
        private static Dictionary<string, string> _contextLogFiles = new Dictionary<string, string>();
        private static Dictionary<string, string> _contextRotatedPaths = new Dictionary<string, string>();

        /// <summary>
        /// Configure the logger
        /// </summary>
        public static void Configure(LogLevel level, string? logFilePath = null, bool consoleOutput = true, LogRotation rotation = LogRotation.NONE)
        {
            _currentLogLevel = level;
            _baseLogFilePath = logFilePath;
            _consoleOutput = consoleOutput;
            _logRotation = rotation;

            // Generate the initial rotated path
            _currentRotatedPath = GetRotatedFilePath();
            _logFilePath = _currentRotatedPath;

            // Create or clear log file if specified
            if (_logFilePath != null)
            {
                try
                {
                    File.WriteAllText(_logFilePath, $"=== Log started at {DateTime.Now:yyyy-MM-dd HH:mm:ss} ===\n");
                }
                catch (Exception ex)
                {
                    Console.Error.WriteLine($"Failed to initialize log file: {ex.Message}");
                }
            }
        }

        /// <summary>
        /// Configure context-specific log files
        /// </summary>
        public static void ConfigureContextLogFiles(Dictionary<string, string> contextLogFiles)
        {
            lock (_lock)
            {
                _contextLogFiles = new Dictionary<string, string>(contextLogFiles);
                _contextRotatedPaths.Clear();

                // Initialize each context log file
                foreach (var kvp in _contextLogFiles)
                {
                    try
                    {
                        var directory = Path.GetDirectoryName(kvp.Value);
                        if (!string.IsNullOrEmpty(directory) && !Directory.Exists(directory))
                        {
                            Directory.CreateDirectory(directory);
                        }
                        File.WriteAllText(kvp.Value, $"=== Log started at {DateTime.Now:yyyy-MM-dd HH:mm:ss} [{kvp.Key}] ===\n");
                        _contextRotatedPaths[kvp.Key] = kvp.Value;
                    }
                    catch (Exception ex)
                    {
                        Console.Error.WriteLine($"Failed to initialize log file for context {kvp.Key}: {ex.Message}");
                    }
                }
            }
        }

        /// <summary>
        /// Set the current log level
        /// </summary>
        public static void SetLogLevel(LogLevel level)
        {
            _currentLogLevel = level;
        }

        /// <summary>
        /// Get the current log level
        /// </summary>
        public static LogLevel GetLogLevel()
        {
            return _currentLogLevel;
        }

        /// <summary>
        /// Log a DEBUG level message
        /// </summary>
        public static void Debug(string message, string? context = null)
        {
            Log(LogLevel.DEBUG, message, context);
        }

        /// <summary>
        /// Log an INFO level message
        /// </summary>
        public static void Info(string message, string? context = null)
        {
            Log(LogLevel.INFO, message, context);
        }

        /// <summary>
        /// Log a WARN level message
        /// </summary>
        public static void Warn(string message, string? context = null)
        {
            Log(LogLevel.WARN, message, context);
        }

        /// <summary>
        /// Log an ERROR level message
        /// </summary>
        public static void Error(string message, string? context = null)
        {
            Log(LogLevel.ERROR, message, context);
        }

        /// <summary>
        /// Generate the rotated file path based on rotation setting
        /// </summary>
        private static string? GetRotatedFilePath()
        {
            if (_baseLogFilePath == null)
            {
                return null;
            }

            if (_logRotation == LogRotation.NONE)
            {
                return _baseLogFilePath;
            }

            var directory = Path.GetDirectoryName(_baseLogFilePath) ?? "";
            var fileNameWithoutExtension = Path.GetFileNameWithoutExtension(_baseLogFilePath);
            var extension = Path.GetExtension(_baseLogFilePath);

            var now = DateTime.Now;
            string suffix;

            if (_logRotation == LogRotation.HOURLY)
            {
                suffix = now.ToString("yyyy-MM-dd_HH");
            }
            else // DAILY
            {
                suffix = now.ToString("yyyy-MM-dd");
            }

            var rotatedFileName = $"{fileNameWithoutExtension}_{suffix}{extension}";
            return Path.Combine(directory, rotatedFileName);
        }

        /// <summary>
        /// Core logging method
        /// </summary>
        private static void Log(LogLevel level, string message, string? context = null)
        {
            if (level < _currentLogLevel)
            {
                return;
            }

            string timestamp = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss.fff");
            string levelStr = level.ToString().PadRight(5);
            string contextStr = context != null ? $"[{context}] " : "";
            string logLine = $"{timestamp} {levelStr} {contextStr}{message}";

            lock (_lock)
            {
                // Console output
                if (_consoleOutput)
                {
                    switch (level)
                    {
                        case LogLevel.DEBUG:
                            Console.ForegroundColor = ConsoleColor.Gray;
                            break;
                        case LogLevel.INFO:
                            Console.ForegroundColor = ConsoleColor.White;
                            break;
                        case LogLevel.WARN:
                            Console.ForegroundColor = ConsoleColor.Yellow;
                            break;
                        case LogLevel.ERROR:
                            Console.ForegroundColor = ConsoleColor.Red;
                            break;
                    }
                    Console.WriteLine(logLine);
                    Console.ResetColor();
                }

                // File output with rotation check
                // First, check if there's a context-specific log file
                if (context != null && _contextLogFiles.ContainsKey(context))
                {
                    try
                    {
                        var contextLogFile = _contextLogFiles[context];
                        File.AppendAllText(contextLogFile, logLine + "\n");
                    }
                    catch
                    {
                        // Silently fail to avoid recursive logging issues
                    }
                }
                else if (_baseLogFilePath != null)
                {
                    try
                    {
                        // Check if we need to rotate to a new file
                        var newRotatedPath = GetRotatedFilePath();
                        if (newRotatedPath != _currentRotatedPath)
                        {
                            _currentRotatedPath = newRotatedPath;
                            _logFilePath = _currentRotatedPath;

                            // Write header to new rotated file
                            if (_logFilePath != null && !File.Exists(_logFilePath))
                            {
                                File.WriteAllText(_logFilePath, $"=== Log started at {DateTime.Now:yyyy-MM-dd HH:mm:ss} ===\n");
                            }
                        }

                        if (_logFilePath != null)
                        {
                            File.AppendAllText(_logFilePath, logLine + "\n");
                        }
                    }
                    catch
                    {
                        // Silently fail to avoid recursive logging issues
                    }
                }
            }
        }

        /// <summary>
        /// Enable logging to file
        /// </summary>
        public static void EnableFileLogging(string filePath)
        {
            _logFilePath = filePath;
            if (_logFilePath != null)
            {
                try
                {
                    File.WriteAllText(_logFilePath, $"=== Log started at {DateTime.Now:yyyy-MM-dd HH:mm:ss} ===\n");
                }
                catch (Exception ex)
                {
                    Console.Error.WriteLine($"Failed to initialize log file: {ex.Message}");
                }
            }
        }

        /// <summary>
        /// Disable logging to file
        /// </summary>
        public static void DisableFileLogging()
        {
            _logFilePath = null;
        }

        /// <summary>
        /// Enable console output
        /// </summary>
        public static void EnableConsoleOutput()
        {
            _consoleOutput = true;
        }

        /// <summary>
        /// Disable console output
        /// </summary>
        public static void DisableConsoleOutput()
        {
            _consoleOutput = false;
        }
    }
}

File: php/Assembler/src/Common/Logger.php
<?php

namespace Assembler\Common;

/**
 * Simple custom logger for consistent logging across all language implementations.
 * Provides file-based logging with different log levels.
 */
class Logger
{
    const DEBUG = 0;
    const INFO = 1;
    const WARN = 2;
    const ERROR = 3;
    const NONE = 4;

    const ROTATION_NONE = 0;
    const ROTATION_HOURLY = 1;
    const ROTATION_DAILY = 2;

    private static $logLevelNames = [
        self::DEBUG => 'DEBUG',
        self::INFO => 'INFO ',
        self::WARN => 'WARN ',
        self::ERROR => 'ERROR',
        self::NONE => 'NONE '
    ];

    private static $logLevelColors = [
        self::DEBUG => "\033[90m",  // Gray
        self::INFO => "\033[97m",   // White
        self::WARN => "\033[93m",   // Yellow
        self::ERROR => "\033[91m"   // Red
    ];

    private static $colorReset = "\033[0m";
    private static $currentLogLevel = self::INFO;
    private static $logFilePath = null;
    private static $consoleOutput = true;
    private static $logRotation = self::ROTATION_HOURLY;
    private static $logsDirectory = null;
    private static $contextLogFiles = [];

    /**
     * Configure the logger (no log file path - use setLogsDirectory instead)
     */
    public static function configure(int $level, bool $consoleOutput = true, int $rotation = self::ROTATION_NONE): void
    {
        self::$currentLogLevel = $level;
        self::$consoleOutput = $consoleOutput;
        self::$logRotation = $rotation;
    }

    /**
     * Set the logs directory - the ONLY way to specify where logs are stored
     */
    public static function setLogsDirectory(string $logsDirectory): void
    {
        self::$logsDirectory = $logsDirectory;
    }

    /**
     * Set the current log level
     */
    public static function setLogLevel(int $level): void
    {
        self::$currentLogLevel = $level;
    }

    /**
     * Get the current log level
     */
    public static function getLogLevel(): int
    {
        return self::$currentLogLevel;
    }

    /**
     * Log a DEBUG level message
     */
    public static function debug(string $message, ?string $context = null): void
    {
        self::log(self::DEBUG, $message, $context);
    }

    /**
     * Log an INFO level message
     */
    public static function info(string $message, ?string $context = null): void
    {
        self::log(self::INFO, $message, $context);
    }

    /**
     * Log a WARN level message
     */
    public static function warn(string $message, ?string $context = null): void
    {
        self::log(self::WARN, $message, $context);
    }

    /**
     * Log an ERROR level message
     */
    public static function error(string $message, ?string $context = null): void
    {
        self::log(self::ERROR, $message, $context);
    }

    /**
     * Core logging method
     */
    private static function log(int $level, string $message, ?string $context = null): void
    {
        if ($level < self::$currentLogLevel) {
            return;
        }

        $timestamp = date('Y-m-d H:i:s.') . substr((string)microtime(), 2, 3);
        $levelStr = self::$logLevelNames[$level];
        $contextStr = $context !== null ? "[{$context}] " : '';
        $logLine = "{$timestamp} {$levelStr} {$contextStr}{$message}";

        // Console output
        if (self::$consoleOutput) {
            $color = self::$logLevelColors[$level] ?? '';
            echo $color . $logLine . self::$colorReset . "\n";
        }

        // File output with rotation check
        // First, check if there's a context-specific log file
        if ($context !== null && isset(self::$contextLogFiles[$context])) {
            $logFile = self::$contextLogFiles[$context];
            $directory = dirname($logFile);

            // Check if directory exists and is writable
            if (is_dir($directory) && is_writable($directory)) {
                try {
                    @file_put_contents($logFile, $logLine . "\n", FILE_APPEND);
                    return; // Don't write to default log file if context file is used
                } catch (\Exception $e) {
                    // Silently fail to avoid recursive logging issues
                }
            }
        }

        if (self::$logFilePath !== null) {
            try {
                $directory = dirname(self::$logFilePath);
                if (is_dir($directory) && is_writable($directory)) {
                    @file_put_contents(self::$logFilePath, $logLine . "\n", FILE_APPEND);
                }
            } catch (\Exception $e) {
                // Silently fail to avoid recursive logging issues
            }
        }
    }

    /**
     * Enable logging to file
     */
    public static function enableFileLogging(string $filePath): void
    {
        self::$logFilePath = $filePath;
        try {
            $timestamp = date('Y-m-d H:i:s');
            file_put_contents($filePath, "=== Log started at {$timestamp} ===\n");
        } catch (\Exception $e) {
            error_log("Failed to initialize log file: " . $e->getMessage());
        }
    }

    /**
     * Disable logging to file
     */
    public static function disableFileLogging(): void
    {
        self::$logFilePath = null;
    }

    /**
     * Enable console output
     */
    public static function enableConsoleOutput(): void
    {
        self::$consoleOutput = true;
    }

    /**
     * Disable console output
     */
    public static function disableConsoleOutput(): void
    {
        self::$consoleOutput = false;
    }

    /**
     * Clear all log files (main and context-specific)
     */
    /**
     * Clear all log files in the logs directory
     */
    public static function clearLogs(): void
    {
        if (self::$logsDirectory === null) {
            return;
        }

        if (is_dir(self::$logsDirectory)) {
            $files = @scandir(self::$logsDirectory);
            if ($files !== false) {
                foreach ($files as $file) {
                    if (pathinfo($file, PATHINFO_EXTENSION) === 'log') {
                        @unlink(self::$logsDirectory . DIRECTORY_SEPARATOR . $file);
                    }
                }
            }
        }

        // Remove the marker file used to detect initialisation so the next request
        // will recreate the log headers.
        $marker = self::$logsDirectory . DIRECTORY_SEPARATOR . '.logs_initialized';
        if (file_exists($marker)) {
            @unlink($marker);
        }
    }

    /**
     * Clear old log files older than the specified number of days
     */
    public static function clearOldLogs(int $days): void
    {
        if (self::$logsDirectory === null) {
            return;
        }

        $cutoffTime = time() - ($days * 24 * 60 * 60);
        
        if (is_dir(self::$logsDirectory)) {
            $files = @scandir(self::$logsDirectory);
            if ($files !== false) {
                foreach ($files as $fileName) {
                    if (pathinfo($fileName, PATHINFO_EXTENSION) !== 'log') {
                        continue;
                    }

                    $fullPath = self::$logsDirectory . DIRECTORY_SEPARATOR . $fileName;
                    if (is_file($fullPath)) {
                        $mtime = @filemtime($fullPath);
                        if ($mtime !== false && $mtime < $cutoffTime) {
                            @unlink($fullPath);
                        }
                    }
                }
            }
        }
    }

    /**
     * Configure context-specific log files (replaces all existing contexts)
     */
    public static function configureContextLogFiles(array $contextLogFiles): void
    {
        self::$contextLogFiles = $contextLogFiles;

        // Initialize each context log file
        foreach ($contextLogFiles as $context => $path) {
            try {
                $directory = dirname($path);
                if (!empty($directory) && !is_dir($directory)) {
                    @mkdir($directory, 0755, true);
                }

                // Check if directory is writable before attempting to write
                if (is_dir($directory) && is_writable($directory)) {
                    // Only write header if file doesn't exist (don't overwrite)
                    if (!file_exists($path)) {
                        $timestamp = date('Y-m-d H:i:s');
                        @file_put_contents($path, "=== Log started at {$timestamp} [{$context}] ===\n");
                    }
                }
            } catch (\Exception $e) {
                // Silently fail to avoid logging errors in production
            }
        }
    }

    /**
     * Add context-specific log files (merges with existing contexts)
     */
    public static function addContextLogFiles(array $contextLogFiles): void
    {
        // Add or update each context
        foreach ($contextLogFiles as $context => $path) {
            // Skip if already configured with same path
            if (isset(self::$contextLogFiles[$context]) && self::$contextLogFiles[$context] === $path) {
                continue;
            }

            try {
                $directory = dirname($path);
                if (!empty($directory) && !is_dir($directory)) {
                    @mkdir($directory, 0755, true);
                }

                // Check if directory is writable before attempting to write
                if (is_dir($directory) && is_writable($directory)) {
                    // Only write header if file doesn't exist (don't overwrite)
                    if (!file_exists($path)) {
                        $timestamp = date('Y-m-d H:i:s');
                        @file_put_contents($path, "=== Log started at {$timestamp} [{$context}] ===\n");
                    }
                }

                self::$contextLogFiles[$context] = $path;
            } catch (\Exception $e) {
                // Silently fail to avoid logging errors in production
            }
        }
    }

    /**
     * Remove specific context log files
     */
    public static function removeContextLogFiles(string ...$contexts): void
    {
        foreach ($contexts as $context) {
            unset(self::$contextLogFiles[$context]);
        }
    }
}
